siege/siege_result_analysis.py

```diff
@@ -0,0 +1,101 @@
+#!/usr/bin/env python3
+#
+# Reads csv files for clean experimental performance and memory which, are
+# generated by running siege multiple times, interleaving clean and
+# experimental. Each perf csv file looks something like this:
+#
+#   Date & Time,         Trans, Elap Time, Data Trans, Resp Time, Trans Rate, Throughput, Concurrent, OKAY,  Failed
+#   2018-10-07 09:40:58, 50000, 4.12,      151,        0.00,      12135.92,   36.65,      45.23,      50000, 0
+#   2018-10-07 09:41:06, 50000, 3.92, 151, 0.00, 12755.10, 38.52, 47.26, 50000, 0
+#   2018-10-07 09:41:13, 50000, 3.47, 151, 0.00, 14409.22, 43.52, 47.62, 50000, 0
+#   2018-10-07 09:41:21, 50000, 3.45, 151, 0.00, 14492.75, 43.77, 45.88, 50000, 0
+#   2018-10-07 09:41:28, 50000, 3.49, 151, 0.00, 14326.65, 43.27, 46.28, 50000, 0
+#
+# And the memory csv files look like:
+#   EnvoyMem, VSZ, RSS
+#   4208832,343468,52508
+#   4220184,343468,55032
+#   4187648,343472,52200
+#   4223848,343468,56532
+#   4187192,343472,54928
+#
+# This script prints out standard deviation for each data-set, to make it
+# obvious if it's noisy, and some basic comparisons.
+#
+# Python3 is required to provide access to the built-in 'statistics' package.
+# If this ever becomes an issue, computing mean/stddev wouldn't be too bad.
+
+import csv
+import statistics
+import sys
+
+# Parses a CSV file into an array of maps from column-name to value.
+# https://stackoverflow.com/questions/21572175/convert-csv-file-to-list-of-dictionaries
+def parseCsv(file):
+  with open(file) as f:
+    return [{k: v for k, v in row.items()}
+            for row in csv.DictReader(f, skipinitialspace=True)]
+
+# Extracts the values of a particular column-name from the array of map,
+# returning the array of values.
+def columnValues(data, column_name):
+  return [float(row[column_name]) for row in data]
+
+def main(argv):
+  if len(argv) != 5:
+    print('Usage: %s clean_perf_csv clean_mem_csv experimental_perf_csv experimental_mem_csv' % argv[0])
+    sys.exit(1)
+
+  # Save the performance CSV data from the clean and experimental runs, using
+  # an array of maps.
+  clean = parseCsv(argv[1])
+  experimental = parseCsv(argv[3])
+
+  # Set up a matrix for printing the analyzed perf results.
+  matrix = [["", "Clean", "Std Dev", "Experimental", "Std Dev", "Improvement"],
+            ["", "-----", "-------", "------------", "-------", "-----------"]]
```

Comment:
Yeah, as long as we have machine generated output as well I think that's good.

What's the opposition to doing a venv setup and using one of the existing CSV table formatter in Python? We already have a bunch of examples of venv in the main repo. IMHO, this is basically Python best practice, since you then make available a ton of others libraries at low cost and avoid future developers of this script from having to manually craft things that are already done elsewhere. The benefits of being standalone are relatively low, since venv/pip are totally standard.

Reply:
No opposition; just trying to be expedient. The 'column' command was revealed to me in the stackoverflow link you shared and was pretty simple.

However when I python-ize the main script I'll switch to using native Python rather than 'column'.